packages/connections/src/XHRConnection.js

```diff
@@ -0,0 +1,206 @@
+
+    /**
+     * @property {object}
+     * @protected
+     * @name XHRConnection#headers
+     */
+    Object.defineProperty(this, "headers", { value: headers });
+
+    /**
+     * @property {string}
+     * @protected
+     * @name XHRConnection#responseType
+     */
+    Object.defineProperty(this, "responseType", { value: responseType });
+
+    /**
+     * @property {XMLHttpRequest}
+     * @protected
+     * @name XHRConnection#xhr
+     */
+    Object.defineProperty(this, "xhr", { value: new XMLHttpRequest() });
+
+    this.xhr.timeout = timeout;
+  }
+
+  get response() {
+    return this.xhr.response;
+  }
+
+  get status() {
+    return this.xhr.status;
+  }
+
+  /**
+   * create, prepare, open and send the xhr request
+   * @param {string} [token] – authentication token
+   */
+  open() {
+    super.open();
+
+    if (this.state !== XHRConnection.INIT) {
+      throw new Error(
+        `Can only open initialized Connections. State is "${this.state}"`
+      );
+    }
+
+    this.xhr.addEventListener("progress", () => {
+      this.emit(
+        ConnectionEvent.DATA,
+        new ConnectionEvent(this, ConnectionEvent.DATA)
+      );
+    });
+
+    this.xhr.addEventListener("load", () => {
+      this.state = XHRConnection.CLOSED;
+
+      if (this.status >= 400) {
+        this.emit(
+          ConnectionEvent.ERROR,
+          new ConnectionEvent(this, ConnectionEvent.ERROR)
+        );
+
+        return;
+      }
+
+      this.emit(
+        ConnectionEvent.COMPLETE,
+        new ConnectionEvent(this, ConnectionEvent.COMPLETE)
+      );
+    });
+
+    this.xhr.addEventListener("abort", () => {
+      this.state = XHRConnection.CLOSED;
+
+      this.emit(
+        ConnectionEvent.ABORT,
+        new ConnectionEvent(this, ConnectionEvent.ABORT)
+      );
+    });
+
+    this.xhr.addEventListener("error", () => {
+      this.state = XHRConnection.CLOSED;
+
+      this.emit(
+        ConnectionEvent.ERROR,
+        new ConnectionEvent(this, ConnectionEvent.ERROR)
+      );
+    });
+
+    this.xhr.addEventListener("timeout", () => {
+      this.state = XHRConnection.CLOSED;
+
+      this.emit(
+        ConnectionEvent.TIMEOUT,
+        new ConnectionEvent(this, ConnectionEvent.TIMEOUT)
+      );
+    });
+
+    this.xhr.open(this.method, this.url);
+    this.state = XHRConnection.OPEN;
+    this.emit(
+      ConnectionEvent.OPEN,
+      new ConnectionEvent(this, ConnectionEvent.OPEN)
+    );
+
+    Object.keys(this.headers).forEach(key => {
+      if (this.headers[key] !== undefined && this.headers[key] !== null) {
+        this.xhr.setRequestHeader(key, this.headers[key]);
+      }
+    });
+
+    this.xhr.responseType = this.responseType;
+    this.xhr.send(this.body);
+  }
+
+  /**
+   * Close the connection
+   * @description Close the connection and abort open requests
+   */
+  close() {
+    super.close();
+
+    if (this.state !== XHRConnection.OPEN) {
+      throw new Error(
```

Comment:
I like what this function did before - silently fail. Since state changes or invocations of close could be sitting on the call stack, the only safe way to ever call close will be to wrap it in a check for `Connection.state === XHRConnection.OPEN` first so we don't accidentally error, so I don't think this makes sense.

Reply:
Yes, thats exactly the idea behind this: to force the client to explicitly check what he is doing. 

If someone wants to deal with the Connections & the CM directly, he should know what he is doing and make sure that everything is handled as expected. Otherwise one should use the `Resource Service` (which is TBD).

Assumptions and silent fails are dangerous in some situations: For example, If `you` create a connection, and pass it to `me` (for whatever reason), I dont know if the connection is already closed or not. If I dont check it, but assume its not, add `on('complete',...)` and call `close()`, then it could happend that my added events are never fired.

But I also agree that we shouldnt worry about these details in later use, thats why we want to wrap all this with the `Resource Service` package.

Reply:
Thanks @juliangieseke, I understand your point about silent fails. I'd argue that `you` in your case above should check the `Connection.state` before you attach your `on('complete',...)` to make sure the connection is not `CLOSED` and if it is, handle accordingly. That is your responsibility to handle this and it isn't forced by throwing an error . Hence I think we should not throw an error here because calling close on a `CLOSED` connection doesn't hurt anybody.

Reply:
Okay, so we want to throw if a component isnt created correctly, but not if the "wrong" method is called. Also I dont want to have silent fails, which means I need to think about this for a second. Maybe just returning true/false or some errorcodes could be enough. :thinking_face: